### PR TITLE
Refactoring tags extensibility

### PR DESCRIPTION
**Pull request by sebastienros**

### CODE REVIEW

**Comment by sebastienros** [comment]
/cc @hishamco @jtkech

I am very happy of these changes. It adds support for `{{-` and `-}}` and won't break on `{%-` and `-%}` as it's the default behavior.
I also completely changed the extensibility story, it's much easier now and we can still create complex grammars. But the standard cases are much more simple.
I am displaying the benchmark in the readme too.

**Comment by hishamco** [comment]
Thanks for this effort @sebastienros, I will have a look to it ASAP, and I will write down my comments if there is

File: Fluid/BaseFluidTemplate.cs [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+namespace Fluid
+{
+    public class BaseFluidTemplate<T> : IFluidTemplate where T : IFluidTemplate, new()
+    {
+        static BaseFluidTemplate()
+        {
+            // Necessary to force the custom template class static constructor
+            // c.f. https://github.com/sebastienros/fluid/issues/19
+            new T();
```
Comment (hishamco) [inline Fluid/BaseFluidTemplate.cs]: I'm still worrying about this line 😟

File: Fluid/BaseFluidTemplate.cs [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        }
+
+        public static FluidParserFactory Factory { get; } = new FluidParserFactory();
+
+        public IList<Statement> Statements { get; set; }
+
+        public BaseFluidTemplate()
+        {
+            Statements = new List<Statement>();
```
Comment (hishamco) [inline Fluid/BaseFluidTemplate.cs]: Can we move this to the property declaration

File: Fluid/Tags/SimpleTag.cs [hunk 3]
```diff
@@ -0,0 +1,23 @@
+﻿using System.IO;
+using System.Text.Encodings.Web;
+using System.Threading.Tasks;
+using Fluid.Ast;
+using Irony.Parsing;
+
+namespace Fluid.Tags
+{
+    public abstract class SimpleTag : ITag
```
Comment (hishamco) [inline Fluid/Tags/SimpleTag.cs]: Could we use `DefaultTag` instead?

**Comment by jtkech** [comment]
Looks great 👍 i'm just starting to look.

File: Fluid.MvcViewEngine/FluidViewTemplate.cs [hunk 4]
```diff
@@ -1,6 +1,15 @@
...
-    public class FluidViewTemplate : FluidTemplate<ActivatorFluidParserFactory<FluidViewParser>>
+    public class FluidViewTemplate : BaseFluidTemplate<FluidViewTemplate>
     {
+        static FluidViewTemplate()
+        {
+            Factory.RegisterTag<LayoutTag>("layout");
+            Factory.RegisterTag<RenderBodyTag>("renderbody");
+            Factory.RegisterBlock<RegisterSectionBlock>("section");
+            Factory.RegisterTag<RenderSectionTag>("rendersection");
+        }
```
Comment (jtkech) [inline Fluid.MvcViewEngine/FluidViewTemplate.cs]: 👍 "je languis" to try it in O2.

File: Fluid.MvcViewEngine/Tags/RegisterSectionBlock.cs [hunk 5]
```diff
@@ -3,24 +3,18 @@
...
 using System.Threading.Tasks;
 using Fluid.Ast;
+using Fluid.Tags;
 
-namespace Fluid.MvcViewEngine.Statements
+namespace Fluid.MvcViewEngine.Tags
 {
-    public class RegisterSectionStatement : TagStatement
+    public class RegisterSectionBlock : IdentifierBlock
     {
```
Comment (jtkech) [inline Fluid.MvcViewEngine/Tags/RegisterSectionBlock.cs]: So, in O2 i will rename all `Statement` with `Tag / Block` ;)

File: Fluid.MvcViewEngine/Tags/RenderSectionTag.cs [hunk 6]
```diff
@@ -3,24 +3,18 @@
...
 using System.Threading.Tasks;
 using Fluid.Ast;
+using Fluid.Tags;
 
-namespace Fluid.MvcViewEngine.Statements
+namespace Fluid.MvcViewEngine.Tags
 {
-    public class RenderSectionStatement : Statement
+    public class RenderSectionTag : IdentifierTag
     {
```
Comment (jtkech) [inline Fluid.MvcViewEngine/Tags/RenderSectionTag.cs]: Oh yes, so there are different kind of `Tag / Block`. So the grammar / parser is updated according to this ...

File: Fluid.Tests/CustomGrammarTests.cs [hunk 7]
```diff
@@ -1,85 +1,188 @@
...
+    }
+
+    public class ShoutTag : ITag
     {
-        protected override Statement BuildTagStatement(ParseTreeNode node)
+        public BnfTerm GetSyntax(FluidGrammar grammar)
         {
-            var tag = node.ChildNodes[0];
+            return grammar.Identifier + grammar.Range;
+        }
```
Comment (jtkech) [inline Fluid.Tests/CustomGrammarTests.cs]: But we can customize the grammar.

File: Fluid.Tests/CustomGrammarTests.cs [hunk 8]
```diff
@@ -1,85 +1,188 @@
...
 
-            switch (tag.Term.Name)
-            {
-                case "yolo":
-                    EnterBlock(tag);
-                    return null;
+        public Statement Parse(ParseTreeNode node, ParserContext context)
+        {
+            var identifier = node.ChildNodes[0].ChildNodes[0].Token.Text;
```
Comment (jtkech) [inline Fluid.Tests/CustomGrammarTests.cs]: And customize the parser.

**Comment by jtkech** [comment]
Wow, LGTM.

Tell me when i can reference a new package version.

**Pull request merged**